modules/rostests/apitests/shlwapi/PathUnExpandEnvStringsForUser.c

```diff
@@ -0,0 +1,122 @@
+/*
+ * PROJECT:         ReactOS api tests
+ * LICENSE:         LGPLv2.1+ - See COPYING.LIB in the top level directory
+ * PURPOSE:         Tests for PathUnExpandEnvStringsForUser
+ * PROGRAMMERS:     Katayama Hirofumi MZ
+ */
+
+#include <apitest.h>
+#include <shlwapi.h>
+#include <stdio.h>
+
+#define DO_TEST(Res, hToken, TestStr, ExpStr, Len) \
+do { \
+    BOOL ret = PathUnExpandEnvStringsForUserW((hToken), (TestStr), OutStr, Len); \
+    ok(ret == (Res), "Tested %s, expected returned value %d, got %d\n", \
+       wine_dbgstr_w((TestStr)), (Res), ret); \
+    if (ret) \
+        ok(_wcsicmp(OutStr, (ExpStr)) == 0, "Tested %s, expected %s, got %s\n", \
+           wine_dbgstr_w((TestStr)), wine_dbgstr_w((ExpStr)), wine_dbgstr_w(OutStr)); \
+} while (0)
+
+// PathUnExpandEnvStringsForUserW
+typedef BOOL (WINAPI *PATHUNEXPANDENVSTRINGSFORUSERW)(HANDLE hToken, LPCWSTR pszPath, LPWSTR pszUnExpanded, INT cchUnExpanded);
+PATHUNEXPANDENVSTRINGSFORUSERW pPathUnExpandEnvStringsForUserW = NULL;
+
+START_TEST(PathUnExpandEnvStringsForUser)
+{
+    DWORD ret;
+    WCHAR OutStr[MAX_PATH], TestStr[MAX_PATH];
+    HINSTANCE hShlwapi;
+
+    hShlwapi = LoadLibraryA("shlwapi");
+    if (hShlwapi == NULL)
+    {
+        skip("shlwapi.dll was not loaded\n");
+        return;
+    }
+
+    pPathUnExpandEnvStringsForUserW =
+        (PATHUNEXPANDENVSTRINGSFORUSERW)GetProcAddress(hShlwapi, "PathUnExpandEnvStringsForUserW");
+
+    if (pPathUnExpandEnvStringsForUserW == NULL)
+    {
+        printf("PathUnExpandEnvStringsForUserW is not public\n");
+        pPathUnExpandEnvStringsForUserW =
+            (PATHUNEXPANDENVSTRINGSFORUSERW)GetProcAddress(hShlwapi, (LPCSTR)(LONG_PTR)466);
+    }
+    if (pPathUnExpandEnvStringsForUserW == NULL)
```

Comment:
ReactOS, in its current state, will not return NULL here, but will raise an exception when the function is called, resulting in a CRASH result in Testman. It would be best to adjust our stub in shell32 at the same time so that it will fail gracefully.
```
466 stdcall -stub -noname PathUnExpandEnvStringsForUserW(ptr wstr ptr long)
```
Should do the trick.

Reply:
Modified shlwapi.spec.